src/rest/class-wp-rest-abilities-run-controller.php

```diff
@@ -0,0 +1,311 @@
+					),
+				),
+
+				// TODO: We register ALLMETHODS because at route registration time, we don't know
+				// which abilities exist or their types (resource vs tool). This is due to WordPress
+				// load order - routes are registered early, before plugins have registered their abilities.
+				// This approach works but could be improved with lazy route registration or a different
+				// architecture that allows type-specific routes after abilities are registered.
+				// This was the same issue that we ended up seeing with the Feature API.
+				array(
+					'methods'             => WP_REST_Server::ALLMETHODS,
+					'callback'            => array( $this, 'run_ability_with_method_check' ),
+					'permission_callback' => array( $this, 'run_ability_permissions_check' ),
+					'args'                => $this->get_run_args(),
+				),
+				'schema' => array( $this, 'get_run_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Executes an ability with HTTP method validation.
+	 *
+	 * @since 0.1.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function run_ability_with_method_check( $request ) {
+		$ability = wp_get_ability( $request['id'] );
+
+		if ( ! $ability ) {
+			return new WP_Error(
+				'rest_ability_not_found',
+				__( 'Ability not found.' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		// Check if the HTTP method matches the ability type
+		$meta = $ability->get_meta();
+		$type = isset( $meta['type'] ) ? $meta['type'] : 'tool';
+		$method = $request->get_method();
+
+		if ( 'resource' === $type && 'GET' !== $method ) {
+			return new WP_Error(
+				'rest_invalid_method',
+				__( 'Resource abilities require GET method.' ),
+				array( 'status' => 405 )
+			);
+		} elseif ( 'resource' !== $type && 'POST' !== $method ) {
+			return new WP_Error(
+				'rest_invalid_method',
+				__( 'Tool abilities require POST method.' ),
+				array( 'status' => 405 )
+			);
+		}
+
+		return $this->run_ability( $request );
+	}
+
+	/**
+	 * Executes an ability.
+	 *
+	 * @since 0.1.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function run_ability( $request ) {
+		$ability = wp_get_ability( $request['id'] );
+
+		if ( ! $ability ) {
+			return new WP_Error(
+				'rest_ability_not_found',
+				__( 'Ability not found.' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		if ( 'GET' === $request->get_method() ) {
+			$input = $request->get_query_params();
+			unset( $input['_locale'], $input['context'], $input['page'], $input['per_page'] );
+		} else {
+			$input = $request->get_json_params() ?: array();
+		}
+
+		$input_validation = $this->validate_input( $ability, $input );
+		if ( is_wp_error( $input_validation ) ) {
+			return $input_validation;
+		}
```

Comment:
This part is not needed as it's covered inside `execute`.

Reply:
I added a TODO for now. The issue is that the false doesn't give us more detailed information about why it failed.

Like you noted,  if we update the main method to return WP_Error, we can use that instead.

I think that might make the most sense, and folks can catch it if needed.